crates/sui-types/src/move_package.rs

```diff
@@ -129,10 +161,65 @@ impl MovePackage {
         Ok(pkg)
     }
 
-    pub fn from_module_iter<T: IntoIterator<Item = CompiledModule>>(
+    /// Create an initial version of the package along with this version's type origin and linkage
+    /// tables.
+    pub fn new_initial(
+        version: SequenceNumber,
+        modules: Vec<CompiledModule>,
+        max_move_package_size: u64,
+        dependencies: &[MovePackage],
+    ) -> Result<Self, ExecutionError> {
+        let type_origin_table = Self::build_initial_type_origin_table(&modules);
+        Self::from_module_iter_with_type_origin_table(
+            version,
+            modules,
+            max_move_package_size,
+            type_origin_table,
+            dependencies,
+        )
+    }
+
+    /// Create an upgraded version of the package along with this version's type origin and linkage
+    /// tables.
+    pub fn new_upgraded(
+        predecessor: &MovePackage,
+        modules: Vec<CompiledModule>,
+        max_move_package_size: u64,
+        dependencies: &[MovePackage],
+    ) -> Result<Self, ExecutionError> {
+        let type_origin_table = Self::build_upgraded_type_origin_table(predecessor, &modules);
+        let mut new_version = SequenceNumber::from(predecessor.version().value());
+        new_version.increment();
+        // TODO: compute all upgraded metadata
+        Self::from_module_iter_with_type_origin_table(
+            new_version,
+            modules,
+            max_move_package_size,
+            type_origin_table,
+            dependencies,
+        )
+    }
+
+    fn get_direct_dependencies(modules: Vec<CompiledModule>) -> BTreeSet<ObjectID> {
+        let mut direct_deps = BTreeSet::new();
+        for m in modules {
+            for dep_handle in m.module_handles() {
+                // exclude the root package
+                if dep_handle != m.self_handle() {
+                    let dep_id: ObjectID = (*m.address_identifier_at(dep_handle.address)).into();
+                    direct_deps.insert(dep_id);
+                }
+            }
+        }
+        direct_deps
+    }
```

Comment:
Nit: This can be a free/top-level function + possible simpler implementation below:

```suggestion
    fn immediate_dependencies(modules: &Vec<CompiledModule>) -> BTreeSet<ObjectID> {
        modules
            .iter()
            .flat_map(|m| m.immediate_dependencies())
            .map(|dep| dep.address().into())
            .collect()
    }
```

Also, it won't be sufficient to just strip out `self_handle` per-module, because modules could depend on each other, within a package, so we should strip out the ID of the package at the end instead (the suggestion above doesn't do that).